### PR TITLE
Add internlm-xcomposer2, forward and sft supported

### PR DESCRIPTION
**Pull request by cocoshe**
对齐 internlm-xcomposer2-7b 的前向，以及sft的step loss和step lr

### CODE REVIEW

**Comment by cocoshe** [comment]
@LokeZhou 辛苦review~

gradient_accumulation_steps为8时

> warmup 阶段step-lr

output_lr


> warmup 阶段step-loss

output_loss

> warmup 之后 step-lr

after_warmup2


> warmup 之后 step-loss

after_warmup1

File: paddlemix/models/internlm_xcomposer2/modeling.py [hunk 1]
```diff
@@ -0,0 +1,1306 @@
...
+                )
+                use_cache = False
+        all_hidden_states = () if output_hidden_states else None
+        all_self_attns = () if output_attentions else None
+        next_decoder_cache = () if use_cache else None
+        for idx, decoder_layer in enumerate(self.layers):
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+            past_key_value = past_key_values[idx] if past_key_values is not None else None
+            if self.gradient_checkpointing and self.training:
```
Comment (cocoshe) [inline paddlemix/models/internlm_xcomposer2/modeling.py:787]: 这部分怎么处理呢？Paddle侧有`gradient_checkpointing`相关的机制嘛？这里涉及了`torch.utils.checkpoint.checkpoint`，不太清楚paddle这部有没有相关实现

File: paddlemix/auto/processing.py [hunk 2]
```diff
@@ -46,11 +46,7 @@ def get_processor_mapping():
             continue
 
         # 2. find the `*processing.py` file as the identifier of ProcessorMixin class,
-        model_name = None
-        if "qwen_vl" in file_name:  # qwen_vl use qwen_vl_processing
-            model_name = "qwen_vl"
-        else:
-            model_name = file_name.split("_")[0]
+        model_name = file_name[:-14]  # remove `.processing.py` suffix
```
Comment (LokeZhou) [inline paddlemix/auto/processing.py:56]: 这个不能去掉，给定14的index对其他模型会有未知错误；要么用关键字模糊匹配给定，要么就直接从file_name参数给定

File: paddlemix/models/internlm_xcomposer2/modeling.py [hunk 3]
```diff
@@ -0,0 +1,1306 @@
...
+                #     def custom_forward(*inputs):
+                #         return module(*inputs, output_attentions, None, im_mask
+                #             )
+                #     return custom_forward
+                # layer_outputs = torch.utils.checkpoint.checkpoint(
+                #     create_custom_forward(decoder_layer), hidden_states,
+                #     attention_mask, position_ids, None)
+                pass
+                # todo: checkpoint api in paddle?
+            else:
```
Comment (cocoshe) [inline paddlemix/models/internlm_xcomposer2/modeling.py]: 这部分怎么处理呢？Paddle侧有`gradient_checkpointing`相关的机制嘛？这里涉及了`torch.utils.checkpoint.checkpoint`，不太清楚paddle这部有没有相关实现

File: paddlemix/examples/internlm_xcomposer2/README.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            [
+                "<img>train/17895/image.png</img>\nWhich is this organism's scientific name?\nCastor canadensis\nNorth American beaver\n",
+                "Castor canadensis"
+            ]
+        ]
+    }
+]
+```
+
+对于带图像输入的内容可表示为 `<img>img_path</img>\n{your prompt}`,"img_path"可以是本地的图片或网络地址。
```
Comment (LokeZhou) [inline paddlemix/examples/internlm_xcomposer2/README.md]: 本地图片或url

File: paddlemix/examples/internlm_xcomposer2/chat_demo.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+)
+args = parser.parse_args()
+
+MODEL_PATH = args.from_pretrained
+# init model and tokenizer
+model = InternLMXComposer2ForCausalLM.from_pretrained(MODEL_PATH).eval()
+tokenizer = InternLMXComposer2Tokenizer.from_pretrained(MODEL_PATH)
+
+text = "<ImageHere>Please describe this image in detail."
+image = "../image1.jpg"
```
Comment (LokeZhou) [inline paddlemix/examples/internlm_xcomposer2/chat_demo.py]: text 和image改为参数传入可以看看llava的

File: paddlemix/models/internlm_xcomposer2/modeling.py [hunk 6]
```diff
@@ -0,0 +1,1306 @@
...
+
+from .configuration import InternLMXcomposer2Config as InternLM2Config
+
+_CONFIG_FOR_DOC = "InternLM2Config"
+logger = logging.get_logger(__name__)
+
+
+def build_vision_tower():
+    vision_tower = "openai/clip-vit-large-patch14-336"
+    return CLIPVisionTower(vision_tower)
```
Comment (LokeZhou) [inline paddlemix/models/internlm_xcomposer2/modeling.py:48]: 这个看能否在config里面定义

File: paddlemix/models/internlm_xcomposer2/modeling.py [hunk 7]
```diff
@@ -0,0 +1,1306 @@
...
+
+
+def build_vision_projector():
+    projector_type = "mlp2x_gelu"
+    mm_hidden_size = 1024
+    hidden_size = 4096
+    mlp_gelu_match = re.match("^mlp(\\d+)x_gelu$", projector_type)
```
Comment (LokeZhou) [inline paddlemix/models/internlm_xcomposer2/modeling.py]: 这些参数都尽量用config.json传入

File: paddlemix/models/internlm_xcomposer2/modeling.py [hunk 8]
```diff
@@ -0,0 +1,1306 @@
...
+        self.select_feature = "patch"
+        self.load_model()
+        self.resize_pos()
+
+    def load_model(self):
+        self.vision_tower = transformers.CLIPVisionModel.from_pretrained(self.vision_tower_name)
+        out_0 = self.vision_tower
+        out_0.stop_gradient = not False
+        self.vision_tower = out_0
+        self.is_loaded = True
```
Comment (LokeZhou) [inline paddlemix/models/internlm_xcomposer2/modeling.py:91]: 这样的写法是否能训练对齐？如果是一个model，可能需要这样https://github.com/PaddlePaddle/PaddleMIX/blob/develop/paddlemix/models/llava/base_model.py#L86

File: paddlemix/models/internlm_xcomposer2/modeling.py [hunk 9]
```diff
@@ -0,0 +1,1306 @@
...
+            )
+            out_1 = paddle.create_parameter(
+                shape=new_pos_embed.to(pos_embed_checkpoint.dtype).shape,
+                dtype=new_pos_embed.to(pos_embed_checkpoint.dtype).numpy().dtype,
+                default_initializer=paddle.nn.initializer.Assign(new_pos_embed.to(pos_embed_checkpoint.dtype)),
+            )
+            out_1.stop_gradient = not True
+            self.vision_tower.vision_model.positional_embedding.weight = out_1
+            self.vision_tower.vision_model.position_ids = paddle.arange(end=new_num).expand(shape=(1, -1))
+            self.is_resize_pos = True
```
Comment (LokeZhou) [inline paddlemix/models/internlm_xcomposer2/modeling.py:127]: 同上

File: paddlemix/models/internlm_xcomposer2/modeling.py (hunk 3 above)
Comment (LokeZhou) [reply paddlemix/models/internlm_xcomposer2/modeling.py]: 可以先直接去掉

**Comment by LokeZhou** [comment]
曲线图是否是减了block或layer数的？如果是，标注一下；另外这个曲线不太平滑，是否torch的也这样？

**Comment by cocoshe** [comment]
主要参数配置：
训练参数 gradient_accumulation_steps=1
组网配置 num_hidden_layers=1, dtype="float32"

+ pytorch 与 paddle 的 step loss

step_loss_accu_1


+ pytorch 与 paddle 的 step lr


step_lr_accu_1

+ pytorch 与 paddle 的 diff loss(差值)

diff_loss_accu_1


+ pytorch 与 paddle 的 diff lr(差值)  （全为0）


diff_lr_accu_1

**Review by LokeZhou — APPROVED** [review]
